File: Data Structures/Queue.py
class Queue:
    def __init__(self):
        self.first = None
        self.last = None

    class Node:
        def __init__(self, value):
            self.value = value
            self.next = None

    def enqueue(self, value):
        node = self.Node(value)
        if self.last is not None:
            self.last.next = node
            self.last = node
        else:
            self.last = node
            self.first = node

    def dequeue(self):
        if self.isEmpty():
            return None
        else:
            val = self.first.value
            self.first = self.first.next
            if self.first is None:
                self.last = None 
            return val


    def peek(self):
        return self.first.value

    def isEmpty(self):
        if self.first is None and self.last is None:
            return True

    def print(self):
        string = ""
        currentNode = self.first
        if currentNode is None:
            return "Empty"
        while currentNode is not None:
            string += str(currentNode.value)
            if currentNode.next is not None:
                string += ", "
            currentNode = currentNode.next
        return string


q = Queue()
print(q.print())

q.enqueue(1)
print(q.print())

q.enqueue(2)
print(q.print())

q.enqueue(3)
print(q.print())

print(q.peek())

print(q.dequeue())
print(q.print())

print(q.dequeue())
print(q.print())

print(q.dequeue())
print(q.print())

print(q.dequeue())
print(q.print())

File: Data Structures/Queue.java
public class Queue<T>{

    Node<T> first;
    Node<T> last;

    public Queue(){
        first = null;
        last = null;
    }

    public class Node<I>{
        I value;
        Node<I> next;
        public Node(I value){
            this.value = value;
        }
    }

    public void enqueue(T value){
        Node<T> node = new Node<T>(value);
        if(isEmpty()){
            first = node;
            last = node;
        }
        else{
            last.next = node;
            last = node;
        }
    }

    public T dequeue(){
        if(isEmpty()){
            return null;
        }
        else{
            T val = first.value;
            first = first.next;
            if (first==null){
                last = null;
            }
            return val;
        }
    }

    public T peek(){
        if(isEmpty()){
            return null;
        }
        else{
            return first.value;
        }
    }

    public boolean isEmpty(){
        return(first == null && last == null);
    }

    public String print(){
        if(isEmpty()){
            return "Empty";
        }
        else{
            StringBuilder sb = new StringBuilder();
            Node<T> currentNode = first;
            while(currentNode != null){
                sb.append(currentNode.value);
                if(currentNode.next != null){
                    sb.append(", ");
                }
                currentNode = currentNode.next;
            }
            return sb.toString();
        }
    }

    public static void main(String[] args){
        Queue<Integer> q = new Queue<Integer>();
        System.out.println(q.print());
        
        q.enqueue(1);
        System.out.println(q.print());
        
        q.enqueue(2);
        System.out.println(q.print());
        
        q.enqueue(3);
        System.out.println(q.print());
        
        System.out.println(q.peek());
        
        System.out.println(q.dequeue());
        System.out.println(q.print());
        
        System.out.println(q.dequeue());
        System.out.println(q.print());
        
        System.out.println(q.dequeue());
        System.out.println(q.print());
        
        System.out.println(q.dequeue());
        System.out.println(q.print());
    }
}